src/main/java/com/google/sps/data/Chat.java

```diff
@@ -58,38 +58,64 @@ public static Chat getChat(Datastore datastore, String chatid) {
     }
   }
 
-  public static List<Chat> getChats(Datastore datastore, String participantid) {
+  public static List<Chat> getUserChats(Datastore datastore, String participantid) {
     
     List<Chat> chats = new ArrayList<>();
-    Query<Entity> chatQuery = Query
-    .newEntityQueryBuilder()
-    .setKind("Chat")
-    .build();
+    Query<Entity> chatQuery = Query.newEntityQueryBuilder().setKind("chat").build();
     QueryResults<Entity> chatEntities = datastore.run(chatQuery);
 
-    // Iterate through the chats
-    // for (Entity entity : chatEntities.asIterable()) {
-    //   long chatid = entity.getKey.getId();
-    //   List<String> participantList = Arrays.asList(entity.getProperty("participants").split(","));
+    while(chatEntities.hasNext()){
+        Entity entity = chatEntities.next();
+        String participants = entity.getString("participants");
+        List<String> participantList = Arrays.asList(participants.split(","));
 
-    //   if (participantList.contains(participantid)){
-    //     Chat chat = new Chat(entity);
-    //     chats.add(chat);
-    //   }
-    // }
+        if (participantList.contains(participantid)){
+            Chat chat = new Chat(entity);
+            chats.add(chat);
+        }
+    }
     return chats;
   }
 
+  public static List<Chat> getAllChats(Datastore datastore) {
+    
+    List<Chat> chats = new ArrayList<>();
+    Query<Entity> chatQuery = Query.newEntityQueryBuilder().setKind("chat").build();
+    QueryResults<Entity> chatEntities = datastore.run(chatQuery);
 
-  long getId(){
+    while(chatEntities.hasNext()){
+        Entity entity = chatEntities.next();
+        Chat chat = new Chat(entity);
+        chats.add(chat);
+    }
+    return chats;
+  }
+
+  public long getId(){
     return id;
   }
 
-  String getParticipants(){
+  public String getParticipants(){
     return participants;
   }
 
-  List<String> getMessages(){
+  public String getMessages(){
     return messages;
   }
+
+  public static boolean chatExists(Datastore datastore, String participantId1, String participantId2){
+    List<Chat> chats = Chat.getAllChats(datastore);
+    ListIterator<Chat> chatsIterator = chats.listIterator();
+    String participants;
+
+    while(chatsIterator.hasNext()){
+        Chat chat = chatsIterator.next();
+        participants = chat.getParticipants();
+        if(participants.contains(participantId1) || participants.contains(participantId2)){
+            return true;
+        }
+    }
+
+    return false;
```

Comment:
I do not recommend fetching all the entities and filtering them like this. Not when Datastore has [query filters](https://cloud.google.com/datastore/docs/concepts/queries#filters) for things like this. Glancing through the docs, it looks like you want to follow the [**Multiple equality filters** example](https://cloud.google.com/datastore/docs/concepts/queries#multiple_equality_filters):

```java
Query<Entity> query = Query.newEntityQueryBuilder()
    .setKind("chat")
    .setFilter(CompositeFilter.and(
        PropertyFilter.eq("participants", participant1), PropertyFilter.eq("participants", participant2)))
    .build();
```

Of course, there are also filters for checking whether a string property contains another string, but I don't recommend the list of participants be stored as a string.

Reply:
Thank you. Id see this, it was confusing seeing the different types of filters especially without seeing any clear examples of how they can be used. Now I see how this one does

Reply:
Im remembering now that the reason why it was difficult finding how to implement a list property is because almost all the resources found like what you did use functions for an outdated version of appengine or datastore, that doesnt work with the entity classes provided by com.google.cloud.datastore - they use com.google.appengine.api

Reply:
Ah yeah, the students last year were using a now-deprecated version of the Datastore API :/